File: 106_ConstructBinaryTreeFromInorderAndPostorder.cpp
// Topic: Construct Binary Tree from Inorder and Postorder
// Description: Given inorder and postorder traversal of a tree, construct the binary tree.
// Approach: Use the last element of postorder as root; split inorder into left/right using index
// map; recurse.

#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* build(vector<int>& inorder, int iL, int iR, vector<int>& postorder, int pL, int pR,
                unordered_map<int, int>& idx)
{
    if (iL > iR || pL > pR)
        return nullptr;
    int rootVal = postorder[pR];
    TreeNode* root = new TreeNode(rootVal);
    int inRootIdx = idx[rootVal];
    int leftSize = inRootIdx - iL;
    root->left = build(inorder, iL, inRootIdx - 1, postorder, pL, pL + leftSize - 1, idx);
    root->right = build(inorder, inRootIdx + 1, iR, postorder, pL + leftSize, pR - 1, idx);
    return root;
}

TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder)
{
    int n = inorder.size();
    unordered_map<int, int> idx;
    for (int i = 0; i < n; ++i) idx[inorder[i]] = i;
    return build(inorder, 0, n - 1, postorder, 0, n - 1, idx);
}

void preorderPrint(TreeNode* root)
{
    if (!root)
        return;
    cout << root->val << ' ';
    preorderPrint(root->left);
    preorderPrint(root->right);
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<pair<vector<int>, vector<int>>> tests = {{{9, 3, 15, 20, 7}, {9, 15, 7, 20, 3}},
                                                    {{2, 1, 3}, {2, 3, 1}}};

    for (int i = 0; i < (int)tests.size(); ++i)
    {
        auto inorder = tests[i].first;
        auto post = tests[i].second;
        TreeNode* root = buildTree(inorder, post);
        cout << "Test " << (i + 1) << ": ";
        preorderPrint(root);
        cout << '\n';
    }
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n) for index map and recursion stack
*/

File: 106_ConstructBinaryTreeFromInorderAndPostorder.py
# Topic: Construct Binary Tree from Inorder and Postorder
# Description: Given inorder and postorder traversal of a tree, construct the binary tree.
# Approach: Use the last element of postorder as root; split inorder into left/right using index map; recurse.

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        idx_map = {val: i for i, val in enumerate(inorder)}

        def build(iL: int, iR: int, pL: int, pR: int) -> Optional[TreeNode]:
            if iL > iR or pL > pR:
                return None
            root_val = postorder[pR]
            root = TreeNode(root_val)
            in_root_idx = idx_map[root_val]
            left_size = in_root_idx - iL
            root.left = build(iL, in_root_idx - 1, pL, pL + left_size - 1)
            root.right = build(in_root_idx + 1, iR, pL + left_size, pR - 1)
            return root

        return build(0, len(inorder) - 1, 0, len(postorder) - 1)


def preorder_print(root: Optional[TreeNode]) -> List[int]:
    if not root:
        return []
    return [root.val] + preorder_print(root.left) + preorder_print(root.right)


if __name__ == "__main__":
    tests = [([9, 3, 15, 20, 7], [9, 15, 7, 20, 3]), ([2, 1, 3], [2, 3, 1])]
    for i, (inorder, postorder) in enumerate(tests):
        root = Solution().buildTree(inorder, postorder)
        print(f"Test {i + 1}: {preorder_print(root)}")

"""
Time Complexity: O(n)
Space Complexity: O(n) for index map and recursion stack
"""
